File: Synch Java/Dispatcher.java

import java.io.BufferedReader;
import java.io.IOException;
import java.util.ArrayList;

/**
 * Classe despachante.
 * Realiza o despacho de trabalhadores para realizar leitura e escrita na base de dados.
 * @author G710
 */
public class Dispatcher extends Thread {
    private static final int MAX_WORKERS = 10, FIRST = 0;
    
    private Worker[] workers;
    private ArrayList<Worker> idleWorkers; // Lista contendo somente os trabalhadores ociosos no momento
    private BufferedReader src; // Fonte das requisições de escrita/leitura
    
    public Dispatcher(Database db, BufferedReader br) {
        src = br;
        workers = new Worker[MAX_WORKERS];
        idleWorkers = new ArrayList<>();
        idleWorkers.ensureCapacity(MAX_WORKERS);
        for (int i = 0; i < MAX_WORKERS; i++) {
            workers[i] = new Worker(db, this);
            idleWorkers.add(workers[i]);
        }
    }
    
    /**
     * Lógica de execução do despachante.
     */
    @Override
    public void run() {
        try {
            String line;
            String split[];
            int pos, val, sleep;
            // Faz a leitura da fonte e processa as requisições para realizar o despacho
            while ((line = src.readLine()) != null) {
                split = line.split(" ");
                pos = Integer.parseInt(split[1]);
                sleep = Integer.parseInt(split[2]);
                val = Integer.parseInt(split[3]);
                System.out.println(split[0].charAt(0)+" "+pos+" "+sleep+" "+val);
                switch (split[0].charAt(0)) {
                    case 'E':
                        dispatchWriter(pos, val, sleep);
                        break;
                    case 'L':
                        dispatchReader(pos, sleep);
                        break;
                }
            }
            System.out.println("End of requests.\nFinishing Execution...");
             // Ao fim das requisições, espera pelo fim das Threads trabalhadoras
            for (int i = 0; i < MAX_WORKERS; i++) {
                workers[i].getThread().join();
            }
        } catch (IOException | InterruptedException ex) {
            System.out.println(ex);
        }
        
    }
    
    /**
     * Despacha um trabalhador para leitura da base de dados.
     * @param position Posição no vetor
     * @param sleepTime Tempo fictício que a thread "demora" para ler
     * @throws InterruptedException 
     */
    private synchronized void dispatchReader(int position, int sleepTime) throws InterruptedException {
        System.out.println("Seeking/Waiting available workers...");
        while (idleWorkers.isEmpty()) { // Espera enquanto não há trabalhadores ociosos
            wait();
        }
        Worker w = idleWorkers.get(FIRST);
        w.startReading(position, sleepTime);
        idleWorkers.remove(w);
    }
    
    /**
     * Despacha um trabalhador para escrita na base de dados.
     * @param position Posição no vetor
     * @param value Valor a ser escrito
     * @param sleepTime Tempo fictício que a thread "demora" para ler
     * @throws InterruptedException 
     */
    private synchronized void dispatchWriter(int position, int value, int sleepTime) throws InterruptedException {
        System.out.println("Seeking/Waiting available workers...");
        while (idleWorkers.isEmpty()) { // Espera enquanto não há trabalhadores ociosos
            wait();
        }
        Worker w = idleWorkers.get(FIRST);
        w.startWriting(position, value, sleepTime);
        idleWorkers.remove(w);
    }
    
    /**
     * Método chamado por uma thread trabalhadora que finalizou sua execução.
     * @param w Trabalhador que finalizou sua thread
     */
    public synchronized void jobsDone(Worker w) {
        idleWorkers.add(w);
        notify(); // Notifica a thread despachante caso ela esteja esperando trabalhadores ociosos.
    }
}

File: Synch C++/Dispatcher.cpp
#include "Dispatcher.h"

DWORD WINAPI startDispatcherThread( LPVOID arg );


Dispatcher::Dispatcher(Database* db, FILE* src) {
  for (size_t i = 0; i < MAX_WORKERS; i++) {
    workers.push_front(new Worker(db, this));  // Inicializa a lista de workers
  }
  this->source = src;

  listSemaphore = CreateSemaphore(NULL, MAX_WORKERS, MAX_WORKERS, NULL);
  listMutex = CreateMutex(NULL, FALSE, NULL);

  if (listSemaphore == NULL)
    ErrorHandler::Error(UNABLE_TO_CREATE_SEMAPHORE);
  if (listMutex == NULL)
    ErrorHandler::Error(UNABLE_TO_CREATE_MUTEX);
}

void Dispatcher::start() {
  thread = CreateThread(NULL, 0,
               startDispatcherThread, // Método inicial da nova thread criada
               this,       // Passa uma referência deste objeto para chamada do método run()
               0,          // Flag de criação padrão
               &threadID); // Armazena ID da thread criada
}

DWORD WINAPI startDispatcherThread( LPVOID  arg ) {
  if (!arg)
    return 0;

  Dispatcher* d = (Dispatcher*)arg;
  d->run();
  return TRUE;
}

/*
* Método com a lógica de execução da Thread despachante,
* Recebe requisições do arquivo para serem processadas.
**/
void Dispatcher::run() {
  char op;
  int position, executionTime, value;
  // Lê requisições do arquivo
  while (fscanf(source, "%c %d %d %d\n", &op, &position, &executionTime, &value) > 2) {
    printf("%c %d %d %d\n", op, position, executionTime, value);
    requestFromSource(op, position, executionTime, value); // Processa requisições recebidas
  }

  // Espera a finalização das threads dos workers
  printf("End of requests.\nFinishing execution...\n");
  while (workers.size() < MAX_WORKERS) {
    Sleep(DISPATCHER_SLEEP_INTERVAL);
  }
}

/*
* Processa as requisições recebidas.
**/
void Dispatcher::requestFromSource(char op, int position, int executionTime, int value) {
  switch(op) {
    case 'L':
      dispatchReader(position, executionTime);
      break;
    case 'E':
      dispatchWriter(position, executionTime, value);
      break;
  }
}

void Dispatcher::dispatchReader(int position, int executionTime) {
  dispatch(position, executionTime, 0, FALSE);
}

void Dispatcher::dispatchWriter(int position, int executionTime, int value) {
  dispatch(position, executionTime, value, TRUE);
}

/*
* Despacha workers para leitura/escrita no banco de dados.
* Mantém a lista de trabalhadores protegida de acesso simultâneo.
**/
void Dispatcher::dispatch(int position, int executionTime, int value, bool writer) {
  DWORD dwWaitSemaphore, dwWaitMutex;

  printf("Seeking/Waiting available workers...\n");
  // Semáforo de controle do número de workers disponíveis
  dwWaitSemaphore = WaitForSingleObject(listSemaphore, INFINITE);
  dwWaitMutex = WaitForSingleObject(listMutex, INFINITE);  // Down no mutex de acesso à lista

  // Delega a requisição ao primeiro worker da lista de disponíveis e remove-o da lista
  Worker* w = workers.front();
  workers.pop_front();
  if (writer) {
    w->startWriting(position, executionTime, value);
  }
  else {
    int target;
    w->startReading(position, executionTime, &target);
  }

  ReleaseMutex(listMutex);
  if (writer)
    printf("Writer dispatched.\n");
  else
    printf("Reader dispatched.\n");
}

/*
* Método usado para receber a notificação de que uma thread trabalhadora
* terminou sua execução. Reinsere o worker ocioso na lista de disponíveis.
**/
void Dispatcher::update(Worker* w) {
  DWORD dwWaitMutex;

  dwWaitMutex = WaitForSingleObject(listMutex, INFINITE);  // Down no mutex de acesso à lista
  workers.push_back(w);
  ReleaseMutex(listMutex);

  ReleaseSemaphore(listSemaphore, 1, NULL);  // Worker adicionado de volta, up no semáforo da lista
}
